### ISSUE TITLE
Capacity FifoHashCache is breached!

### ISSUE THREAD

**Issue by gjeee** [iotaledger/iri#565]
TipsViewModel  uses 2 FifoHashCache objects called tips and solidTips with a capacity of 5000 each. So the total tipcount should be never more than 10K. However my own node shows a tipcount (solid+non-solid) of more than 10000. A lot of other nodes as well as can be seen on https://iotanode.host

So somehow the add-routine of [FifoHashCache](https://github.com/iotaledger/iri/blob/6433963b4b59e61fce093e308e5271b50d110469/src/main/java/com/iota/iri/controllers/TipsViewModel.java) is not working as intended.

**Comment by GalRogozinski** [comment]
Also `com.iota.iri.controllers.TipsViewModel.FifoHashCache#addAll` is no good since it ignore `capacity`. This isn't the problem because no one is calling it, but it also should be fixed in this issue.

**Comment by gjeee** [comment]
Indeed i also noticed that #addAll issue. Apart from this i have been experimenting with ListOrderedSet (from org.apache.commons.collections4) instead of LinkedHashSet. Using this datastructure did not breach the capacity. Moreover you can randomly access elements since it has a get(index) method which is very useful for  the getRandomSolidTipHash() routine. However this routine does throw now and then IndexOutOfBounds exceptions (caught in "Tip Solidity Rescan" thread) but i have not been able yet to track down why. There are locks in place so it is not a thread thing (????).

**Comment by GalRogozinski** [comment]
Did you understand why the change from LinkedHashSet solves the capacity issue?
It does seem like all methods are synchronized as far as I can tell.

**Comment by GalRogozinski** [comment]
It is apparent that Hash#Equals() and Hash#HashCode() are broken...

**Comment by gjeee** [comment]
if the Hash#Equals() is broken basically means the hashset can contain duplicates....but how does that explain that capacity is breached? it should still contain  <=5000

**Comment by gjeee** [comment]
somehow i am now unable to reproduce the findings. maybe it only happens in a relatively high-load environment.

**Comment by gjeee** [comment]
![image](https://user-images.githubusercontent.com/13155183/37003653-1ca9954e-20ce-11e8-8d0b-695d4a2a2743.png)

i was able to reproduce findings again!!! see attached image: the left-number is size non-solid tips, right number is total tips. 

it happens with dev branch + patch 470 . Probably patch 470 triggers the assertion in Hash#Equals() .

i will try once more with just the dev branch.

**Comment by GalRogozinski** [comment]
Interesting, if you think it is the assertion then apply [patch 571](https://github.com/iotaledger/iri/pull/571).

**Comment by gjeee** [comment]
ok i did some more tests. i can confirm that the breach happens for all the three setups below:

- dev branch + patch 470
- dev
- dev + path 571

so unfortunately the equals fix did not solve it. i will try to check once more whether ListOrderedSet (instead of LinkedHashSet) fixes it.

**Comment by GalRogozinski** [comment]
Since I see you are working on it I assigned this to you

**Comment by gjeee** [comment]
As found previously ListOrderedSet does not exceed capacity. Since this structure uses only equals as comparator and LinkedHashSet both equals && hashCode, could it be that hashCode instead is broken.....i can not find a direct flaw/bug but maybe u know???????

**Comment by footloosejava** [comment]
!!!!  MAYBE THIS IS THE PROBLEM  !!!!!!

I don't think I need to explain this any more.

### Suppose we introduce a Reentrant lock and test access to the methods ...

```
    @Override
    public boolean equals(final Object obj) {
        assert obj instanceof Hash;
        if (obj == null) return false;
        return Arrays.equals(bytes(), ((Hash) obj).bytes());
    }

    @Override
    public int hashCode() {
        if (bytes == null) {
            bytes();
        }
        int temp = hashCode;
        if (reentrantLock.tryLock()) {
            try {
                return temp;
            } finally {
                reentrantLock.unlock();
            }
        } else {
            log.error("Unsynchronized access to hashcode! hashcode(" + temp + ") " +
                    "identityHashCode(" + System.identityHashCode(this) + ")");
            return temp;
        }
    }

    @Override
    public String toString() {
        return Converter.trytes(trits());
    }

    public byte[] bytes() {
        if (bytes == null) {
            if (reentrantLock.tryLock()) {
                bytes = new byte[SIZE_IN_BYTES];
                try {
                    Converter.bytes(trits, 0, bytes, 0, trits.length);
                    hashCode = Arrays.hashCode(this.bytes);
                } finally {
                    reentrantLock.unlock();
                }
            } else {
                log.error("Unsynchronized access to byte array creation! " +
                        "identityHCode(" + System.identityHashCode(this) + ")");
            }
        }
        return bytes;
    }

```

### WHAT WE GET IS THIS ... 

> /08 04:38:12.228 [pool-2-thread-3] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1816607978)
> 03/08 04:38:27.270 [pool-2-thread-11] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(116962895)
> 03/08 04:38:31.867 [pool-2-thread-2] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(139371443)
> 03/08 04:38:37.173 [pool-2-thread-10] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(80349811)
> 03/08 04:38:47.658 [pool-2-thread-11] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(201383231)
> 03/08 04:38:50.370 [pool-2-thread-9] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1406814251)
> 03/08 04:38:55.111 [pool-2-thread-13] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1342260431)
> 03/08 04:38:55.568 [pool-2-thread-7] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(2030078572)
> 03/08 04:38:59.628 [pool-2-thread-18] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(733284945)
> 03/08 04:39:01.774 [pool-2-thread-11] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1224884955)
> 03/08 04:39:06.856 [pool-2-thread-19] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1640829048)
> 03/08 04:39:08.148 [pool-2-thread-20] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(615810473)
> 03/08 04:39:19.967 [pool-2-thread-18] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(628289879)
> 03/08 04:39:20.042 [pool-2-thread-16] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(864767654)
> 03/08 04:39:20.260 [pool-2-thread-8] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(792887509)
> 03/08 04:39:20.918 [pool-2-thread-1] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1329638591)
> 03/08 04:39:31.299 [pool-2-thread-17] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1296410720)
> 03/08 04:39:35.389 [pool-2-thread-6] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(597167293)
> 03/08 04:39:42.510 [pool-2-thread-10] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(16096818)
> 03/08 04:39:46.607 [pool-2-thread-2] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(193893750)
> 03/08 04:39:50.765 [pool-2-thread-21] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1178736493)
> 03/08 04:39:58.126 [pool-2-thread-20] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(712212303)
> 03/08 04:39:59.735 [pool-2-thread-14] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1349946640)
> 03/08 04:40:01.546 [pool-2-thread-9] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(2097068242)
> 03/08 04:40:05.796 [pool-2-thread-19] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1862988337)
> 03/08 04:40:10.113 [pool-2-thread-16] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1028433575)
> 03/08 04:40:13.334 [pool-2-thread-17] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1132613560)
> 03/08 04:40:16.453 [pool-2-thread-22] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1585644419)
> 03/08 04:40:22.642 [pool-2-thread-20] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(776793817)
> 03/08 04:40:22.680 [pool-2-thread-19] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1798667016)
> 03/08 04:40:26.461 [pool-2-thread-6] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1418940830)
> 03/08 04:40:27.650 [pool-2-thread-10] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1173815368)
> 03/08 04:40:28.492 [pool-2-thread-21] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1287289885)
> 03/08 04:40:29.682 [pool-2-thread-8] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(729616869)
> 03/08 04:40:32.539 [pool-2-thread-12] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(2025075459)
> 03/08 04:40:34.547 [pool-2-thread-17] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1203618576)
> 03/08 04:40:35.599 [pool-2-thread-21] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(163113966)
> 03/08 04:40:37.793 [pool-2-thread-8] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1853852398)
> 03/08 04:40:38.022 [pool-2-thread-7] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(2009557820)
> 03/08 04:40:40.683 [pool-2-thread-1] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1250190887)
> 03/08 04:40:41.950 [pool-2-thread-17] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(46089603)
> 03/08 04:40:42.719 [pool-2-thread-22] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1346692643)
> 03/08 04:40:45.779 [pool-2-thread-8] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1680086118)
> 03/08 04:40:47.770 [pool-2-thread-22] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1417373939)
> 03/08 04:40:49.797 [pool-2-thread-14] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1414075334)
> 03/08 04:40:56.017 [pool-2-thread-13] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1590079679)
> 03/08 04:41:01.973 [pool-2-thread-9] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(733470090)
> 03/08 04:41:05.061 [pool-2-thread-13] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1526280127)
> 03/08 04:41:11.018 [pool-2-thread-22] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(293202968)
> 03/08 04:41:11.067 [pool-2-thread-21] ERROR com.iota.iri.network.Node - Unsynchronized access to byte array creation! identityHC(1595374795)

**Comment by footloosejava** [comment]
📃 Time for some writing 📃

###  .... because also ...

`03/08 05:38:33.176 [pool-2-thread-3] ERROR com.iota.iri.network.Node - Unsynchronised HASHCODE access! hc(0) SystemHC(1555271274)
`

* The same `Hash` object instance is being initialized (bytes created) by more than one thread at the same time.

* Different threads are testing `equals` on the same `Hash` instance while other threads are creating them.

* Different threads are testing `hashcode` while other threads are either checking `equals` and/or making the initialization bytes.

### This is a bad race condition.

It looks like more than thread is trying to instantiate the same object, or maybe even treating it as a different object and trying to use it for itself.

The operating system places operations in the same method out-of-order. making it possible for `bytes==null` to test `false`, while the `hashcode` has not been created and the `bytes` are still being written, resulting in hashcode being returned and used in places such as HashMap and HashSet. Never mind all the `equals` tests on their own.

Since `Hash` is the bricks in `Hashes`, which is further subclassed into `Address`, `Approvee`, `Bundle` and `Tag`, this makes them all very fragile.

I'm done.

**Comment by gjeee** [comment]
this is a major bug-find! well done!

**Comment by gjeee** [comment]
I changed the Hash implementation. I made sure that all the constructors read all the state vars in 1 go in a synchronised fashion. I also synchronised the functions:

- toString()
- bytes()
- trits()
- hashCode()
- trailingZeros() 
- equals()

Let me know what you think. I can do a pull request later.

```

package com.iota.iri.model;

import com.iota.iri.hash.Curl;
import com.iota.iri.hash.Sponge;
import com.iota.iri.hash.SpongeFactory;
import com.iota.iri.storage.Indexable;
import com.iota.iri.utils.Converter;

import java.io.Serializable;
import java.util.Arrays;
import java.util.Objects;
 

public class Hash implements Serializable, Indexable {

    public static final int SIZE_IN_TRITS = 243;
    public static final int SIZE_IN_BYTES = 49;

    public static final Hash NULL_HASH = new Hash(new int[Curl.HASH_LENGTH]);

    private byte[] bytes;
    private int[] trits;
    private int hashCode;
	
    public final Object sync = new Object();

    
    // constructors' bill

    public Hash(){}

    public Hash(final byte[] bytes, final int offset, final int size) {
        fullRead(bytes, offset, size);
    }
    
    public Hash(final int[] trits, final int offset) {
        fullRead(trits, offset);
    }
    
    public Hash(final byte[] bytes) {
        fullRead(bytes, 0, SIZE_IN_BYTES);
    }

    public Hash(final int[] trits) {
        fullRead(trits, 0);
    }

    public Hash(final String trytes) {
        int[] trits = new int[SIZE_IN_TRITS];
        Converter.trits(trytes,trits, 0);
        fullRead(trits, 0);
    }
	
	private void fullRead(final byte[] bytes, final int offset, final int size){
		synchronized (sync) {
			this.bytes = new byte[SIZE_IN_BYTES];
			System.arraycopy(bytes, offset, this.bytes, 0, size - offset > bytes.length ? bytes.length-offset: size);
			this.trits = new int[Curl.HASH_LENGTH];
                        Converter.getTrits(this.bytes, this.trits);
			this.hashCode = Arrays.hashCode(this.bytes);
		}
	}
	
	private void fullRead(final int[] trits, final int offset){
		synchronized (sync) {
			this.trits = new int[SIZE_IN_TRITS];
			System.arraycopy(trits, offset, this.trits, 0, SIZE_IN_TRITS);
			this.bytes = new byte[SIZE_IN_BYTES];
                        Converter.bytes(this.trits, 0, this.bytes, 0, this.trits.length);  
			this.hashCode = Arrays.hashCode(this.bytes);
		}
	}
      

    //
    /*
    public static Hash calculate(byte[] bytes) {
        return calculate(bytes, SIZE_IN_TRITS, new Curl());
    }
    */
    public static Hash calculate(SpongeFactory.Mode mode, int[] trits) {
        return calculate(trits, 0, trits.length, SpongeFactory.create(mode));
    }

    public static Hash calculate(byte[] bytes, int tritsLength, final Sponge curl) {
        int[] trits = new int[tritsLength];
        Converter.getTrits(bytes, trits);
        return calculate(trits, 0, tritsLength, curl);
    }
    public static Hash calculate(final int[] tritsToCalculate, int offset, int length, final Sponge curl) {
        int[] hashTrits = new int[SIZE_IN_TRITS];
        curl.reset();
        curl.absorb(tritsToCalculate, offset, length);
        curl.squeeze(hashTrits, 0, SIZE_IN_TRITS);
        return new Hash(hashTrits);
    }

    public int trailingZeros() {
		synchronized (sync) {
			int index, zeros;
			index = SIZE_IN_TRITS;
			zeros = 0;
			while(index-- > 0 && this.trits[index] == 0) {
				zeros++;
			}
			return zeros;
		}
    }

  

    @Override
    public boolean equals(Object o) {
		synchronized (sync) {
			if (this == o) {
				return true;
			}
			if (o == null || getClass() != o.getClass()) {
				return false;
			}
			Hash hash = (Hash) o;
			return Arrays.equals(this.bytes, hash.bytes());
		}
     }

    @Override
    public int hashCode() {  
		synchronized (sync) {    
			return this.hashCode;
		}
    }

    @Override
    public String toString() {
		synchronized (sync) {
			return Converter.trytes(this.trits);
		}
    }
    
    public byte[] bytes() { 
		synchronized (sync) {    
			return this.bytes;
		}
    }
	
    public int[] trits() {
		synchronized (sync) {
			return this.trits;
		}
    }

    @Override
    public void read(byte[] bytes) {
         fullRead(bytes, 0, SIZE_IN_BYTES);
    }

    @Override
    public Indexable incremented() {
        return null;
    }

    @Override
    public Indexable decremented() {
        return null;
    }

    @Override
    public int compareTo(Indexable indexable) {
        Hash hash = new Hash(indexable.bytes());
        if (this.equals(hash)) {
            return 0;
        }
        long diff = Converter.longValue(hash.trits(), 0, SIZE_IN_TRITS) - Converter.longValue(trits(), 0, SIZE_IN_TRITS);
        if (Math.abs(diff) > Integer.MAX_VALUE) {
            return diff > 0L ? Integer.MAX_VALUE : Integer.MIN_VALUE + 1;
        }
        return (int) diff;
    }
}


```

**Comment by footloosejava** [comment]
Since this is used ... everywhere ... and in all Sets and HashMaps ... better not to synchronize the whole thing like the fix above (although that should work).

If you can give me a day, I have a fix that avoids all the synchronized blocks. I've got it partially coded.

**Comment by footloosejava** [comment]
`compareTo`  can also be changed to check to see if the `Indexable` passed in is also a `Hash`.  Recreating a new hash in ever ordered collection insert ... could create a lot of unnecessary objects when it calls `compareTo`. It could be changed to this.

```
@Override
    public int compareTo(Indexable indexable) {
        Hash hash = (indexable instanceof Hash) ? (Hash) indexable : new Hash(indexable.bytes());
        if (this.equals(hash)) {
            return 0;
        }
        long diff = Converter.longValue(hash.trits(), 0, SIZE_IN_TRITS) - Converter.longValue(trits(), 0, SIZE_IN_TRITS);
        if (Math.abs(diff) > Integer.MAX_VALUE) {
            return diff > 0L ? Integer.MAX_VALUE : Integer.MIN_VALUE + 1;
        }
        return (int) diff;
    }
```

**Comment by gjeee** [comment]
you want to keep the lazy initialization AND no sync blocks???? i know sync blocks are expensive but can you come up with a fix without any sync blocks ????

**Comment by footloosejava** [comment]
You can do it using this:

```
    // THE VARS IN INNER CLASS CANNOT BE ACCESSED
    // UNTIL THE OBJECT IS INSTANTIATED
    // THIS IS THREAD SAFE
    private final class DataSafe {
        private final byte[] bytes;
        private final Integer hashcode;

        private DataSafe(byte[] bytes, Integer hashcode) {
            this.bytes = Objects.requireNonNull(bytes);
            this.hashcode = Objects.requireNonNull(hashcode);
        }
    }
```

Then ... 

```
    public byte[] bytes() {
        DataSafe safe = bytesSafe;
        if (safe == null) {
            synchronized (lock) {
                if (bytesSafe == null) {
                    byte[] temp = new byte[SIZE_IN_BYTES];
                    Converter.bytes(initTrits, 0, temp, 0, initTrits.length);
                    bytesSafe = new DataSafe(temp, Arrays.hashCode(temp));
                }
                safe = bytesSafe;
            }
        }
        return safe.bytes;
    }

    @Override
    public int hashCode() {
        if (bytesSafe == null) {
            bytes();
        }
        return bytesSafe.hashcode;
    }
```

So everywhere you need byte[] and hashcode, check if bytesafe is null, if it is, then call bytes() and
then return the value. This can get rid of the sync blocks with null checks.

You can also make a tristSafe and then check it the same way when trits is called.

The DataSafe need to be a class inside Hash.

That will work. So will synchronized blocks too ... just might make a performance hit in some cases.

I suppose the sync blocks are easier to code ... and maybe even clearer. Given that everything is made of Hash objects (Hashes and subclcasses are collections of them), I'd try to avoid the sync blocks. If there is no performance hit ... well then keep it simple.

Just an idea for your consideration.

**Comment by gjeee** [comment]
thanks this seems like a nice solution... can you either paste the full code of your Hash implementation here or do a commit?? i like to know what @GalRogozinski  thinks as well. i also think we should try to merge this asap to see how the network will  behave after this fix. i think now a lot of illegal/invalid tips are floating around.

**Comment by footloosejava** [comment]
No one has addressed this https://github.com/iotaledger/iri/issues/564
That is very pressing. The TCP neighbor response is fubar because of it ... The queued packets are all sharing the exact same data array -ie. the info is completely redundant in the entire TCP Neighbor queue!

For this, I'll do a fork and submit a patch ASAP. For TCPNeighbor, I just want to make sure we all agree on the problem and maybe someone there can patch it (my big report shows the fix).

**Comment by gjeee** [comment]
unfortunately i am not really proficient in low-level networking stuff.....so pls do what u got to do....i can not help you here....

luckily i only have UDP neighbors :-)

**Comment by footloosejava** [comment]
See https://github.com/iotaledger/iri/pull/577

**Comment by gjeee** [comment]
i applied PR#577 and it looks good so far.....my synchronized version worked as well but i think your implementation is more efficient although it did not do any benchmark-experiments (in terms of timing or resource usage)

**Comment by footloosejava** [comment]
Sounds good! I tried one that didn't do lazy for the trits array and it failed a test due to memory (Heap?). So I guess that is one advantage of lazy init of the arrays.

**Comment by alon-e** [comment]
@footloosejava / @galrogo - can this be closed after #597 was merged?

**Comment by footloosejava** [comment]
@alon-e  @galrogo Closing this is OK with me.

**Comment by GalRogozinski** [comment]
Fixed in #597
